juju/juju
# [JUJU-2222] Support multi-homing on OpenStack

A long time ago, a configuration article, _network_, was introduced for OpenStack models. This has been used since to indicate the internal network that Juju should connect provisioned VMs to. It is required to be set when the OpenStack project used for the model has multiple such networks available.

This is a contrivance that works outside the preferred Juju techniques for choosing networks to connect to. For workloads and machines we have constraints and bindings; for bootstrapping, we should use placement directives.

This patch removes the requirement to configure a single internal OpenStack network for VM connectivity. It upholds the following functionality regarding the _network_ configuration item.
- The option still works and is honoured, preserving the behaviour of current deployments.
- The option can be now set with a comma-delimited list of networks in order that multiple networks are considered.
- The option can be omitted in order that all networks available to the model's OpenStack project are considered.

We will likely consider removal of this configuration article in a future major release.

## QA steps

Requires an OpenStack cloud with multiple networks and subnets available. The author was able to add new networks and subnets to both _ServerStack_ and _CanoniStack_ projects.
- Bootstrap, specifying a single value for _network_ configuration.
- Check that the controller has a single ens/eth device (disregard Fan) in the configured network.
- Check that `juju subnets -m controller` only lists subnets in the configured network.
- Add a model _without_ network configuration.
- `juju add-machine` and check that `juju show-machine x` shows a NIC in each available network.
- `juju remove-machine x` and wait for it to be gone.
- `juju add-space beta <CIDR>` using the CIDR from the network that the controller model is configured for.
- `juju deploy postgresql --bind beta` and check that `juju show-machine x` shows a single NIC in the beta subnet.
- `juju add-machine --constraints spaces=alpha,beta` and check that `juju show-machine x` shows a NIC in each network.
- `juju add-machine --constraints spaces=beta` and check that `juju show-machine x` shows a single NIC in the beta subnet.
- Add a model with multiple configured networks, for example `--config network=some-net-1,some-net-2`.
- Verify the same steps as for the network-less model above.

## Documentation changes

Likely that we'll need to update docs regarding configuring OpenStack models.

## Bug reference

N/A
